drivers/sensor/default_rtio_sensor.c

```diff
@@ -251,66 +251,205 @@ void sensor_processing_with_callback(struct rtio *ctx, sensor_processing_callbac
 {
-	struct sensor_data_generic_header *header = (struct sensor_data_generic_header *)buffer;
+	size_t count = 0;
 
-	ARG_UNUSED(channel_type);
-	*shift = header->shift;
-	return 0;
+	for (size_t i = 0; i < header->num_channels; ++i) {
+		if (channel != header->channels[i]) {
+			continue;
+		}
+		if (count == channel_idx) {
+			*out = values[i];
+			return 0;
+		}
+		++count;
+	}
+	return -EINVAL;
+}
+
+static int decode_three_axis(const struct sensor_data_generic_header *header, const q31_t *values,
+			     struct sensor_three_axis_data *data_out, enum sensor_channel x,
+			     enum sensor_channel y, enum sensor_channel z, size_t channel_idx)
+{
+	int rc;
+
+	data_out->header.base_timestamp_ns = header->timestamp_ns;
+	data_out->header.reading_count = 1;
+	data_out->shift = header->shift;
+	data_out->readings[0].timestamp_delta = 0;
+
+	rc = get_q31_value(header, values, x, channel_idx, &data_out->readings[0].values[0]);
+	if (rc < 0) {
+		return rc;
+	}
+	rc = get_q31_value(header, values, y, channel_idx, &data_out->readings[0].values[1]);
+	if (rc < 0) {
+		return rc;
+	}
+	rc = get_q31_value(header, values, z, channel_idx, &data_out->readings[0].values[2]);
+	if (rc < 0) {
+		return rc;
+	}
+	return 1;
+}
+
+static int decode_q31(const struct sensor_data_generic_header *header, const q31_t *values,
+		      struct sensor_q31_data *data_out, enum sensor_channel channel,
+		      size_t channel_idx)
+{
+	int rc;
+
+	data_out->header.base_timestamp_ns = header->timestamp_ns;
+	data_out->header.reading_count = 1;
+	data_out->shift = header->shift;
+	data_out->readings[0].timestamp_delta = 0;
+
+	rc = get_q31_value(header, values, channel, channel_idx, &data_out->readings[0].value);
+	if (rc < 0) {
+		return rc;
+	}
+	return 1;
 }
 
 /**
- * @brief Default decoder decode N samples
+ * @brief Decode up to N samples from the buffer
  *
- * Decode up to N samples starting at the provided @p fit and @p cit. The appropriate channel types
- * and q31 values will be placed in @p values and @p channels respectively.
+ * This function will never wrap frames. If 1 channel is available in the current frame and
+ * @p max_count is 2, only 1 channel will be decoded and the frame iterator will be modified
+ * so that the next call to decode will begin at the next frame.
  *
- * @param[in]     buffer The data buffer to decode
- * @param[in,out] fit The starting frame iterator
- * @param[in,out] cit The starting channel iterator
- * @param[out]    channels The decoded channel types
- * @param[out]    values The decoded q31 values
- * @param[in]     max_count The maximum number of values to decode
- * @return > 0 The number of decoded values
- * @return 0 Nothing else to decode on this @p buffer
- * @return < 0 Error
+ * @param[in]     buffer The buffer provided on the :c:struct:`rtio` context
+ * @param[in]     channel The channel to decode
+ * @param[in]     channel_idx The index of the channel
+ * @param[in,out] fit The current frame iterator
```

Comment:
Is this per channel (type,idx) or is this the overall frame? When does this get incremented and by how much?

If each channel can produce differing numbers of samples it would seem to be troublesome potentially?

Reply:
So the thinking is that this is basically how we wanted to present the data in the subsystem, but just moved a layer down. This way people can start off using the async kernel drivers and migrate to the subsystem with minimal changes. Having said that, here's how it works (in my mind):

You get back a buffer with data. Lets assume from a FIFO just to make the example more complex (also builds on your question in https://github.com/zephyrproject-rtos/zephyr/pull/61022#pullrequestreview-1622683370).

So you poke the decoder and see that there are:
* 3 frames of `SENSOR_CHAN_ACCEL_XYZ`
* 3 frames of `SENSOR_CHAN_GYRO_XYZ`
* 2 frames of `SENSOR_CHAN_DIE_TEMP`

Lets assume you have enough memory to decode everything at the same time. What you'll get is a decoded data structure for a 3 axes on the accel, it would have a header with a base timestamp and 3 samples (each with an offset). Lets make up the data:
```
header:
  base_timestamp_ns: 500
  reading_count: 3
shift: 4 // scales +/- 16 m/s2
readings:
  - timestamp_delta: 0
    x: 0
    y: 0
    z: 9.8
  - timestamp_delta: 1e9
    x: 0
    y: 0
    z: 9.801
  - timestamp_delta: 2e9
    x: 0
    y: 0
    z: 9.709
```

The gyro would have data that looks similar, but the die temperature would only have 2 frames and would map to a 1d struct (`sensor_q31_data`). It would look like:
```
header:
  base_timestamp_ns: 500
  reading_count: 2
shift: 5 // scales +/- 32 C
readings:
  - timestamp_delta: 0
    value: 23.3
  - timestamp_delta: 2e9
    value:23.35
```

Here you can see, the second reading has a timestamp (`header.base_timestamp_ns + readings[1].timestamp_delta`) that matches the timestamp of the 3rd acceleration reading. So they can be associated by keeping an index. It would look like:
```
decode the 3 channels
is timestamp of all 3 channels the same?
yes
process
increment the readings index for all 3 samples

is timestamp of all 3 channels the same?
no, die temperature is bigger
process accel/gyro combo (maybe calculating orientation)
increment just the accel/gyro reading index

is timestamp of all 3 channel readings the same?
yes
process
```

The same would work if we wanted to only decode 1 frame at a time.

Reply:
What happens when you decode less than the entire fifo and the channels have varying numbers of samples?

Reply:
I think I see the disconnect, the idea is that now there's a frame iterator per channel. So it would look like:

```c
sensor_frame_iterator_t accel_fit = 0, gyro_fit = 0;

decoder->decode(buf, SENSOR_CHAN_ACCEL_XYZ, &accel_fit, 1, accel_output_buffer);
decoder->decode(buf, SENSOR_CHAN_GYRO_XYZ, &gyro_fit, 1, gyro_output_buffer);

accel_ts = accel_output_buffer->header.base_timestamp_ns + accel_output_buffer->readings[0].timestamp_delta;
gyro_ts = gyro_output_buffer->header.base_timestamp_ns + gyro_output_buffer->readings[0].timestamp_delta;

if (accel_ts == gyro_ts) {
  // same frame
} else if (accel_ts < gyro_ts) {
  // accel is from an earlier frame than the gyro
} else {
  // accel is from a later frame than the gyro 
}
```

Reply:
It might make sense then to shove the channel type, channel index, and frame index into a struct.

This way its maybe clearer what you are iterating over.

Something like...

```c
struct sensor_frame_iterator gyro_iter = SENSOR_FRAME_ITER_INIT(decoder, buf, SENSOR_CHAN_GYRO_XYZ, 1);
struct sensor_frame_iterator accel_iter = SENSOR_FRAME_ITER_INIT(decoder, buf, SENSOR_CHAN_ACCEL_XYZ, 1);

sensor_decode(&gyro_iter, gyro_out_data);
sensor_decode(&accel_iter, accel_out_data);
```